File: app/src/main/java/com/example/miniapp/viewmodels/LoginViewModel.java
package com.example.miniapp.viewmodels;

import android.util.Log;

import com.example.miniapp.helper_classes.IPublisher;
import com.example.miniapp.helper_classes.ISubscriber;
import com.example.miniapp.models.ILoginDBManager;

public class LoginViewModel implements IViewModel, IPublisher<Integer> {
    private ILoginDBManager dbManager;
    private ISubscriber<Integer> loginView;

    public LoginViewModel(ISubscriber<Integer> loginView, ILoginDBManager dbManager){
        this.dbManager = dbManager;

        // publish changes to MainActivity (login)
        this.addSub(loginView);
    }

    @Override
    public void openDB() {
        dbManager.openDB();
    }

    @Override
    public void closeDB() {
        dbManager.closeDB();
    }

    public void verify(String email, String password) {
        boolean isEmailRegistered = dbManager.isEmailRegistered(email);
        Boolean isPasswordCorrect = null;
        Log.v("MY TAG", "email reg: " + isEmailRegistered);
        if (isEmailRegistered){
            isPasswordCorrect = dbManager.verifyCredentials(email, password);
            Log.v("MY TAG", "password correct: " + isPasswordCorrect);
        }

        // the following integers are used for the login status
        // 0: email is not registered
        // 1: email is registered, password is correct
        // -1: email is registered, password is incorrect
        if (!isEmailRegistered) {
            notifySubs(0);
        } else {
            if (isPasswordCorrect) {
                notifySubs(1);
            } else {
                notifySubs(-1);
            }
        }
    }

    public void register(String email, String password) {
        dbManager.register(email, password);
    }

    @Override
    public void addSub(ISubscriber<Integer> subscriber) {
        loginView = subscriber;
    }

    @Override
    public void removeSub(ISubscriber<Integer> subscriber) {
    }

    @Override
    public void notifySubs(Integer loginStatus) {
        loginView.update(loginStatus);
    }

}

File: app/src/main/java/com/example/miniapp/viewmodels/LoginViewModel.kt
package com.example.miniapp.viewmodels

import com.example.miniapp.helper_classes.IPublisher
import com.example.miniapp.helper_classes.ISubscriber
import com.example.miniapp.helper_classes.log
import com.example.miniapp.models.ILoginDBManager

class LoginViewModel(loginView: ISubscriber<Int>, private val dbManager: ILoginDBManager) : IPublisher<Int> {
    private lateinit var loginView: ISubscriber<Int>

    init {
        // publish changes to MainActivity (login)
        this.addSub(loginView)
    }

    fun verify(email: String, password: String) {
        val isEmailRegistered = dbManager.isEmailRegistered(email)
        val isPasswordCorrect: Boolean? = if(isEmailRegistered) dbManager.verifyCredentials(email, password) else null

        // the following integers are used for the login status
        // 0: email is not registered
        // 1: email is registered, password is correct
        // -1: email is registered, password is incorrect
        if (!isEmailRegistered) {
            log("register email $email")
            customNotify { loginView.update(0) }
        } else {
            if (isPasswordCorrect!!) {
                log("password correct for $email")
                customNotify { loginView.update(1) }
            } else {
                log("password incorrect for $email")
                customNotify { loginView.update(-1) }
            }
        }
    }

    fun register(email: String, password: String) {
        dbManager.register(email, password)
    }

    override fun addSub(subscriber: ISubscriber<Int>) {
        loginView = subscriber
    }
}